src/pages/resultsView/survival/SurvivalChart.tsx

```diff
@@ -128,9 +135,67 @@ export default class SurvivalChart extends React.Component<ISurvivalChartProps,
         let configurableOpts: StyleOpts = _.merge({}, this.styleOptsDefaultProps, this.props.styleOpts);
         configurableOpts.padding.right = this.props.showLegend ? 300 : configurableOpts.padding.right;
         configurableOpts.legend.x = configurableOpts.width - configurableOpts.padding.right;
+        this.downSamplingDenominatorX = configurableOpts.width - configurableOpts.padding.left - configurableOpts.padding.right;
+        this.downSamplingDenominatorY = configurableOpts.height - configurableOpts.padding.top - configurableOpts.padding.bottom;
         return configurableOpts;
     }
 
+    @computed
+    get allScatterData() {
+        let a = {
+            altered: {
+                numOfCases: this.sortedAlteredPatientSurvivals.length,
+                line: getLineData(this.sortedAlteredPatientSurvivals, this.alteredEstimates),
+                scatterWithOpacity: getScatterDataWithOpacity(this.sortedAlteredPatientSurvivals, this.alteredEstimates),
+                scatter: getScatterData(this.sortedAlteredPatientSurvivals, this.alteredEstimates)
+            },
+            unaltered: {
+                numOfCases: this.sortedUnalteredPatientSurvivals.length,
+                line: getLineData(this.sortedUnalteredPatientSurvivals, this.unalteredEstimates),
+                scatterWithOpacity: getScatterDataWithOpacity(this.sortedUnalteredPatientSurvivals, this.unalteredEstimates),
+                scatter: getScatterData(this.sortedUnalteredPatientSurvivals, this.unalteredEstimates)
+            }
+        };
+        return a;
+    }
+
+    // Only recalculate the scatter data based on the plot filter.
+    // The filter is only available when user zooms in the plot.
+    @computed
+    get scatterData() {
+        let filteredData = _.cloneDeep(this.allScatterData);
```

Comment:
Can you move this whole calculation to SurvivalUtil, and also add unit tests for new methods in SurvivalUtil?

Reply:
Sure.